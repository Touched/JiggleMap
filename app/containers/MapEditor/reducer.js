/*
 *
 * MapEditor reducer
 *
 */

import R from 'ramda';
import { combineReducers } from 'redux';
import undoable, { includeAction } from 'redux-undo';
import * as THREE from 'three';
import { calculateBoundingRectangle } from 'components/Renderer/utils';
import { getToolById } from './utils';

import {
  LOAD_MAIN_MAP,
  LOAD_CONNECTED_MAP,
  EDIT_MAP,
  COMMIT_MAP_EDIT,
  SET_CAMERA_POSITION,
  MOVE_CONNECTION,
  COMMIT_CONNECTION_MOVE,
  RESIZE_VIEWPORT,
  SET_CURRENT_BLOCK,
  MAP_LOADED,
  MOVE_ENTITY,
  COMMIT_ENTITY_MOVE,
  SET_ACTIVE_LAYER,
} from './constants';

const mapData = {
  dimensions: [0, 0],
  height: [],
  collision: [],
  block: [],
};

const blocksetData = {
  id: null,
  palette: [],
  blocks: [],
  tiles: [],
};

const initialDataState = {
  border: mapData,

  // When the user is editing (i.e. dragging the mouse around), the changes they make are
  // applied to the data in 'map' for UI feedback. They are then copied to 'canonicalMap'
  // when they lift the mouse button.
  map: mapData,
  canonicalMap: mapData,
  blocksets: {
    primary: blocksetData,
    secondary: blocksetData,
  },
  scripts: [],
  connections: [],
  canonicalConnectionOffsets: [],
  entities: [],
  canonicalEntityCoordinates: [],
};

const initialEditingState = {
  loading: false,
  linked: false,
  viewportSize: {
    width: 0,
    height: 0,
  },
  camera: {
    x: 0,
    y: 0,
    z: 1,
  },
  activeLayer: 'map',
  activeTool: 'line-tool',
};

function loadMapData(data) {
  const {
    width: dataWidth,
    height: dataHeight,
    data: { block, collision, height },
  } = data;

  return {
    dimensions: [dataWidth, dataHeight],
    block,
    collision,
    height,
  };
}

function loadBlocksetData(data) {
  return {
    primary: {
      id: data.primary.resource.meta.id,
      palette: data.primary.resource.data.palette,
      blocks: data.primary.resource.data.blocks,
      tiles: data.primary.tiles,
    },
    secondary: {
      id: data.secondary.resource.meta.id,
      palette: data.secondary.resource.data.palette,
      blocks: data.secondary.resource.data.blocks,
      tiles: data.secondary.tiles,
    },
  };
}

function setMapDataValue(oldData, newData, index, value) {
  if (value === undefined || oldData[index] === value) {
    return newData;
  }

  const dataCopy = newData || oldData.slice();
  dataCopy[index] = value;
  return dataCopy;
}

function editMap(data, patch) {
  let mapBlockData = null;
  const [width, height] = data.dimensions;

  patch.forEach(({ x, y, block }) => {
    if (x >= 0 && x < width && y >= 0 && y < height) {
      const index = x + (y * width);
      mapBlockData = setMapDataValue(data.block, mapBlockData, index, block);
    }
  });

  if (mapBlockData) {
    return {
      ...data,
      block: mapBlockData,
    };
  }

  return data;
}

export function mapDataReducer(state = initialDataState, action) {
  switch (action.type) {
    case LOAD_MAIN_MAP: {
      let mapState = {};
      if (action.map) {
        const loadedData = loadMapData(action.map.data.map);

        mapState = {
          map: loadedData,
          canonicalMap: loadedData,
          entities: action.map.data.entities,
          canonicalEntityCoordinates: action.map.data.entities.map(({ x, y }) => ({ x, y })),
        };
      }

      const blocksetState = action.blocksets ? { blocksets: loadBlocksetData(action.blocksets) } : {};

      return {
        ...state,
        ...mapState,
        ...blocksetState,
      };
    }
    case LOAD_CONNECTED_MAP:
      return {
        ...state,
        connections: [
          ...state.connections, {
            map: loadMapData(action.map.data.map),
            blocksets: loadBlocksetData(action.blocksets),
            direction: action.direction,
            offset: action.offset,
          },
        ],
        canonicalConnectionOffsets: [
          ...state.canonicalConnectionOffsets,
          action.offset,
        ],
      };
    case EDIT_MAP:
      return {
        ...state,
        // Changes are applied to the canonical map, but saved in the map so that they are
        // visible to the user.
        map: editMap(state.canonicalMap, action.patch),
      };
    case COMMIT_MAP_EDIT:
      return {
        ...state,
        canonicalMap: state.map,
      };
    case MOVE_CONNECTION: {
      const { direction } = state.connections[action.connection];
      const canonicalOffset = state.canonicalConnectionOffsets[action.connection];

      const [width, height] = state.map.dimensions;

      let offset;
      switch (direction) {
        case 'left':
        case 'right':
          offset = Math.max(Math.min(canonicalOffset + action.y, height), -height);
          break;
        case 'up':
        case 'down':
          offset = Math.max(Math.min(canonicalOffset + action.x, width), -width);
          break;
        default:
          offset = 0;
          break;
      }

      const lens = R.lensPath(['connections', action.connection, 'offset']);
      return R.set(lens, offset, state);
    }
    case COMMIT_CONNECTION_MOVE: {
      const lens = R.lensPath(['canonicalConnectionOffsets', action.connection]);
      return R.set(lens, state.connections[action.connection].offset, state);
    }
    case MOVE_ENTITY: {
      const index = state.entities.findIndex(({ id }) => id === action.id);

      if (index >= 0) {
        const [width, height] = state.map.dimensions;
        const { x: canonicalX, y: canonicalY } = state.canonicalEntityCoordinates[index];

        const x = Math.max(Math.min(canonicalX + action.x, width - 1), 0);
        const y = Math.max(Math.min(canonicalY + action.y, height - 1), 0);

        const lens = R.lensPath(['entities', index]);
        return R.over(lens, R.merge(R.__, { x, y }), state); // eslint-disable-line no-underscore-dangle
      }

      return state;
    }
    case COMMIT_ENTITY_MOVE: {
      const index = state.entities.findIndex(({ id }) => id === action.id);

      if (index >= 0) {
        const { x, y } = state.entities[index];
        const lens = R.lensPath(['canonicalEntityCoordinates', index]);
        return R.over(lens, R.merge(R.__, { x, y }), state); // eslint-disable-line no-underscore-dangle
      }

      return state;
    }
    default:
      return state;
  }
}

export function mapEditingReducer(state = initialEditingState, action) {
  switch (action.type) {
    case LOAD_MAIN_MAP: {
      const { viewportSize: { width, height } } = state;
      const { width: mapWidth, height: mapHeight } = action.map.data.map;
      const boundingBox = calculateBoundingRectangle(width, height, mapWidth * 16, mapHeight * 16, 0, 0);
      const min = new THREE.Vector3(0, 0, 0);
      const max = new THREE.Vector3(boundingBox.width, boundingBox.height, 0);
      const box = new THREE.Box3(min, max);
      const midpoint = box.getCenter();

      return {
        ...state,
        camera: {
          ...state.camera,
          x: midpoint.x,
          y: -midpoint.y,
        },
      };
    }
    case SET_CAMERA_POSITION:
      return {
        ...state,
        camera: {
          x: action.x,
          y: action.y,
          z: action.z,
        },
      };
    case RESIZE_VIEWPORT:
      return {
        ...state,
        viewportSize: {
          width: action.width,
          height: action.height,
        },
      };
    case SET_CURRENT_BLOCK:
      return {
        ...state,
        toolState: {
          ...state.toolState,
          currentBlock: action.block,
        },
      };
    case SET_ACTIVE_LAYER:
      return {
        ...state,
        activeLayer: action.layer,
      };
    default: {
      const tool = getToolById(state.activeTool);

      return {
        ...state,
        toolState: tool.reducer(state.toolState, action),
      };
    }
  }
}

// Split the reducers into undoable map data and other editing state
export default combineReducers({
  editing: mapEditingReducer,
  data: undoable(mapDataReducer, {
    limit: 10,
    filter: includeAction([MAP_LOADED, COMMIT_CONNECTION_MOVE, COMMIT_MAP_EDIT, COMMIT_ENTITY_MOVE]),
    ignoreInitialState: true,
  }),
});
