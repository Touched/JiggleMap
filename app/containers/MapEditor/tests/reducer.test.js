import {
  loadMainMap,
  loadConnectedMap,
  editMap,
  commitMapEdit,
  setCameraPosition,
  moveConnection,
  commitConnectionMove,
  moveEntity,
  commitEntityMove,
  setActiveLayer,
} from '../actions';
import { mapDataReducer, mapEditingReducer } from '../reducer';

describe('mapDataReducer', () => {
  it('returns the initial state', () => {
    expect(mapDataReducer(undefined, {})).toMatchSnapshot();
  });

  const mapData = {
    width: 2,
    height: 2,
    data: {
      height: [1, 1, 2, 3],
      collision: [1, 1, 0, 1],
      block: [1, 2, 3, 4],
    },
  };

  const blocksetData = {
    primary: {
      resource: {
        meta: {
          id: 'blockset1',
        },
        data: {
          palette: [[[0, 0, 0, 0]]],
          blocks: [{
            tiles: [{ tile: 0, flipX: false, flipY: false, palette: 0 }],
          }],
        },
      },
      tiles: [1, 2, 3],
    },
    secondary: {
      resource: {
        meta: {
          id: 'blockset2',
        },
        data: {
          palette: [[[1, 2, 3, 4]]],
          blocks: [{
            tiles: [{ tile: 0, flipX: false, flipY: false, palette: 0 }],
          }],
        },
      },
      tiles: [4, 5, 6],
    },
  };

  const loadedMapData = {
    height: [1, 1, 2, 3],
    collision: [1, 1, 0, 1],
    block: [1, 2, 3, 4],
    dimensions: [2, 2],
  };

  const loadedBlocksetData = {
    primary: {
      id: 'blockset1',
      palette: [[[0, 0, 0, 0]]],
      blocks: [{
        tiles: [{ tile: 0, flipX: false, flipY: false, palette: 0 }],
      }],
      tiles: [1, 2, 3],
    },
    secondary: {
      id: 'blockset2',
      palette: [[[1, 2, 3, 4]]],
      blocks: [{
        tiles: [{ tile: 0, flipX: false, flipY: false, palette: 0 }],
      }],
      tiles: [4, 5, 6],
    },
  };

  describe('LOAD_MAIN_MAP', () => {
    it('loads map data and entities into the main map', () => {
      const action = loadMainMap({
        map: {
          data: {
            map: mapData,
            entities: [{
              type: 'unknown',
              x: 0,
              y: 1,
              z: 3,
            }],
          },
        },
      });

      const newState = mapDataReducer(undefined, action);
      expect(newState).toMatchObject({
        map: loadedMapData,
        canonicalMap: loadedMapData,
        entities: [{
          type: 'unknown',
          x: 0,
          y: 1,
          z: 3,
        }],
        canonicalEntityCoordinates: [{
          x: 0,
          y: 1,
        }],
      });
    });

    it('loads the blockset data', () => {
      const action = loadMainMap({
        blocksets: blocksetData,
      });

      const newState = mapDataReducer(undefined, action);
      expect(newState).toMatchObject({
        blocksets: loadedBlocksetData,
      });
    });
  });

  describe('LOAD_CONNECTED_MAP', () => {
    it('adds a connection to the list of connections', () => {
      const action = loadConnectedMap({
        map: {
          data: {
            map: mapData,
          },
        },
        blocksets: blocksetData,
        offset: 3,
        direction: 'up',
      });

      const newState = mapDataReducer(undefined, action);
      expect(newState.connections).toEqual([{
        map: loadedMapData,
        blocksets: loadedBlocksetData,
        offset: 3,
        direction: 'up',
      }]);

      expect(newState.canonicalConnectionOffsets).toEqual([
        3,
      ]);
    });
  });

  describe('EDIT_MAP', () => {
    it('applies a patch', () => {
      const action = editMap([{
        x: 0,
        y: 1,
        block: 1,
      }, {
        x: 1,
        y: 1,
        block: 1,
      }, {
        x: 2,
        y: 1,
        block: 1,
      }]);

      const newState = mapDataReducer({
        canonicalMap: {
          dimensions: [4, 4],
          block: [
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0,
          ],
        },
      }, action);

      expect(newState.map.block).toEqual([
        0, 0, 0, 0,
        1, 1, 1, 0,
        0, 0, 0, 0,
        0, 0, 0, 0,
      ]);
    });
  });

  describe('COMMIT_MAP_EDIT', () => {
    it('updates canonical map', () => {
      const action = commitMapEdit();
      const newState = mapDataReducer({ map: loadedMapData }, action);
      expect(newState.canonicalMap).toEqual(loadedMapData);
    });
  });

  describe('MOVE_CONNECTION', () => {
    it('moves the offset horizontally for a "down" connection', () => {
      const action = moveConnection(0, 10, 3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'down' }],
        canonicalConnectionOffsets: [7],
      }, action);

      expect(newState.connections[0].offset).toEqual(17);
    });

    it('moves the offset horizontally for a "up" connection', () => {
      const action = moveConnection(0, -10, 3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'up' }],
        canonicalConnectionOffsets: [7],
      }, action);

      expect(newState.connections[0].offset).toEqual(-3);
    });

    it('moves the offset vertically for a "left" connection', () => {
      const action = moveConnection(0, 10, 3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'left' }],
        canonicalConnectionOffsets: [7],
      }, action);

      expect(newState.connections[0].offset).toEqual(10);
    });

    it('moves the offset vertically for a "right" connection', () => {
      const action = moveConnection(0, 10, -3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'right' }],
        canonicalConnectionOffsets: [7],
      }, action);

      expect(newState.connections[0].offset).toEqual(4);
    });

    it('ignores non-directional connections', () => {
      const action = moveConnection(0, 10, -3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'dive' }],
        canonicalConnectionOffsets: [0],
      }, action);

      expect(newState.connections[0].offset).toEqual(0);
    });

    it('cannot move a connection out of bounds', () => {
      const action = moveConnection(0, 100, 3);

      const newState = mapDataReducer({
        map: { dimensions: [30, 30] },
        connections: [{ direction: 'down' }],
        canonicalConnectionOffsets: [7],
      }, action);

      expect(newState.connections[0].offset).toEqual(30);
    });
  });

  describe('COMMIT_CONNECTION_MOVE', () => {
    it('updates the canonical offset', () => {
      const action = commitConnectionMove(0);

      const newState = mapDataReducer({
        connections: [{ offset: 7 }],
        canonicalConnectionOffsets: [9],
      }, action);

      expect(newState.canonicalConnectionOffsets[0]).toEqual(7);
    });
  });

  describe('MOVE_ENTITY', () => {
    const state = {
      map: { dimensions: [30, 30] },
      entities: [{
        id: 'entity0',
        x: 0,
        y: 0,
      }],
      canonicalEntityCoordinates: [{
        x: 8,
        y: 9,
      }],
    };

    it('moves the entity', () => {
      const action = moveEntity('entity0', 5, 7);

      const newState = mapDataReducer(state, action);

      expect(newState.entities[0]).toEqual({
        id: 'entity0',
        x: 13,
        y: 16,
      });
    });

    it('ignores unknown entities', () => {
      const action = moveEntity('entity1', 5, 7);

      const newState = mapDataReducer(state, action);

      expect(newState.entities[0]).toEqual({
        id: 'entity0',
        x: 0,
        y: 0,
      });
    });


    it('cannot move an entity out of bounds', () => {
      const action = moveEntity('entity0', 50, 70);

      const newState = mapDataReducer(state, action);

      expect(newState.entities[0]).toEqual({
        id: 'entity0',
        x: 29,
        y: 29,
      });
    });
  });

  describe('COMMIT_ENTITY_MOVE', () => {
    const state = {
      map: { dimensions: [30, 30] },
      entities: [{
        id: 'entity0',
        x: 0,
        y: 0,
      }],
      canonicalEntityCoordinates: [{
        x: 8,
        y: 9,
      }],
    };

    it('updates the canonical coordinates', () => {
      const action = commitEntityMove('entity0');
      const newState = mapDataReducer(state, action);

      expect(newState.canonicalEntityCoordinates[0]).toEqual({
        x: 0,
        y: 0,
      });
    });

    it('ignores unknown entity ids', () => {
      const action = commitEntityMove('entity1');
      const newState = mapDataReducer(state, action);

      expect(newState.canonicalEntityCoordinates[0]).toEqual({
        x: 8,
        y: 9,
      });
    });
  });
});

describe('mapEditingReducer', () => {
  it('returns the initial state', () => {
    expect(mapEditingReducer(undefined, {})).toMatchSnapshot();
  });

  describe('LOAD_MAIN_MAP', () => {
    it('recenters the map', () => {
      const state = {
        viewportSize: {
          width: 500,
          height: 600,
        },
      };

      const action = loadMainMap({
        map: {
          data: {
            map: {
              width: 24,
              height: 20,
            },
          },
        },
      });

      const newState = mapEditingReducer(state, action);

      expect(newState.camera).toMatchSnapshot();
    });
  });

  describe('SET_CAMERA_POSITION', () => {
    it('changes the camera position', () => {
      const action = setCameraPosition(1, 2, 3);

      const newState = mapEditingReducer(undefined, action);
      expect(newState.camera).toEqual({
        x: 1,
        y: 2,
        z: 3,
      });
    });
  });

  describe('SET_ACTIVE_LAYER', () => {
    it('sets the active layer', () => {
      const action = setActiveLayer('collision');
      const newState = mapEditingReducer(undefined, action);
      expect(newState.activeLayer).toEqual('collision');
    });
  });
});
