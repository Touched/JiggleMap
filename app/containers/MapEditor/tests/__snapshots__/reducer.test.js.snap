// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDataReducer returns the initial state 1`] = `
Object {
  "blocksets": Object {
    "primary": Object {
      "blocks": Array [],
      "id": null,
      "palette": Array [],
      "tiles": Array [],
    },
    "secondary": Object {
      "blocks": Array [],
      "id": null,
      "palette": Array [],
      "tiles": Array [],
    },
  },
  "border": Object {
    "block": Array [],
    "collision": Array [],
    "dimensions": Array [
      0,
      0,
    ],
    "height": Array [],
  },
  "canonicalConnectionOffsets": Array [],
  "canonicalEntityCoordinates": Array [],
  "canonicalMap": Object {
    "block": Array [],
    "collision": Array [],
    "dimensions": Array [
      0,
      0,
    ],
    "height": Array [],
  },
  "connections": Array [],
  "entities": Array [],
  "map": Object {
    "block": Array [],
    "collision": Array [],
    "dimensions": Array [
      0,
      0,
    ],
    "height": Array [],
  },
  "scripts": Array [],
}
`;

exports[`mapEditingReducer LOAD_MAIN_MAP recenters the map 1`] = `
Object {
  "x": 192,
  "y": -160,
}
`;

exports[`mapEditingReducer returns the initial state 1`] = `
Object {
  "activeLayer": "map",
  "activeTool": Object {
    "collision": undefined,
    "entities": undefined,
    "height": undefined,
    "map": "drag-connection-tool",
  },
  "camera": Object {
    "x": 0,
    "y": 0,
    "z": 1,
  },
  "linked": false,
  "loading": false,
  "toolState": Object {
    "drag-connection-tool": Object {
      "mouse": Object {
        "clientOffset": null,
        "currentBlock": 0,
        "currentPosition": null,
        "object": null,
        "startingPosition": null,
      },
      "tool": Object {},
    },
  },
}
`;
